PhysicsTools/PatAlgos/src/MuonMvaEstimator.cc

```diff
@@ -0,0 +1,139 @@
+  tmvaReader_.AddVariable("LepGood_sip3d",                 &sip_              );
+  tmvaReader_.AddVariable("log(abs(LepGood_dxy))",         &log_abs_dxyBS_    ); 
+  tmvaReader_.AddVariable("log(abs(LepGood_dz))",          &log_abs_dzPV_     );
+  tmvaReader_.AddVariable("LepGood_segmentCompatibility",  &segmentCompatibility_);
+  tmvaReader_.BookMVA("BDTG",weightsfile);
+  dRmax_ = dRmax;
+  initialized_ = true;
+};
+
+float ptRel(const reco::Candidate::LorentzVector& muP4, 
+	    const reco::Candidate::LorentzVector& jetP4, 
+	    bool subtractMuon=true) 
+{
+  reco::Candidate::LorentzVector jp4 = jetP4;
+  if (subtractMuon) jp4-=muP4;
+  float dot = muP4.Vect().Dot( jp4.Vect() );
+  float ptrel = muP4.P2() - dot*dot/jp4.P2();
+  ptrel = ptrel>0 ? sqrt(ptrel) : 0.0;
+  return ptrel;
+}
+
+void MuonMvaEstimator::computeMva(const pat::Muon& muon,
+				  const reco::Vertex& vertex,
+				  const reco::JetTagCollection& bTags,
+				  const reco::JetCorrector* correctorL1,
+				  const reco::JetCorrector* correctorL1L2L3Res)
+{
+  if (not initialized_) 
+    throw cms::Exception("FatalError") << "MuonMVA is not initialized";
+  pt_                   = muon.pt();
+  eta_                  = muon.eta();
+  segmentCompatibility_ = muon.segmentCompatibility();
+  miniRelIsoCharged_ = muon.miniPFIsolation().chargedHadronIso();
+  miniRelIsoNeutral_ = muon.miniPFIsolation().neutralHadronIso();
+
+  double dB2D  = muon.dB(pat::Muon::BS2D);
+  double dB3D  = muon.dB(pat::Muon::PV3D);
+  double edB3D = muon.edB(pat::Muon::PV3D);
+  sip_  = edB3D>0?fabs(dB3D/edB3D):0.0; 
+  log_abs_dxyBS_     = log(fabs(dB2D)); 
+  log_abs_dzPV_      = log(fabs(muon.muonBestTrack()->dz(vertex.position())));
+
+  // compute jet-related information
+  // jetPtRel_          = jetPtRel;
+  // jetPtRatio_        = std::min(jetPtRatio,1.5);
+  // jetBTagCSV_        = std::max(jetBTagCSV,0.0);
+  // jetNDauCharged_    = jetNDauCharged;
+
+  //Initialise loop variables
+  double minDr = 9999;
+  double jecL1L2L3Res = 1.;
+  double jecL1 = 1.;
+
+  jetPtRatio_ = -99;
+  jetPtRel_   = -99;
+  jetBTagCSV_ = -999;
+  jetNDauCharged_ = -1;
+
+  for (const auto& tagI: bTags){
+    // for each muon with the lepton 
+    double dr = deltaR(*(tagI.first), muon);
+    if(dr > minDr) continue;  
+    minDr = dr;
+      
+    const reco::Candidate::LorentzVector& muP4(muon.p4()); 
+    reco::Candidate::LorentzVector jetP4(tagI.first->p4());
+
+    if (correctorL1 && correctorL1L2L3Res){
+      jecL1L2L3Res = correctorL1L2L3Res->correction(*(tagI.first));
+      jecL1 = correctorL1->correction(*(tagI.first));
+    }
+
+    // Get b-jet info
+    jetBTagCSV_ = tagI.second;
+    jetNDauCharged_ = 0;
+    for (auto jet: tagI.first->getJetConstituentsQuick()){
+      const reco::PFCandidate *pfcand = dynamic_cast<const reco::PFCandidate*>(jet);
+      if (pfcand==nullptr) continue;
+      if (pfcand->charge()==0) continue;
+      if (!pfcand->bestTrack()) continue;
```

Comment:
```auto bestTrackPtr = pfcand->bestTrack(); ``` and then operate on bestTrackPtr below so that the code to derive the best track pointer is not called repeatedly (it's not a trivial operation).

Reply:
fixed